File: 1219-path-with-maximum-gold/1219-path-with-maximum-gold.py
class Solution:
    def getMaximumGold(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        
        direc = [0, 1, 0, -1, 0]
        
        def dfs(r, c):
            if r < 0 or r >= n or c < 0 or c >= m or grid[r][c] <= 0:
                return 0
            # visited node value is reversed
            grid[r][c] *= -1
            mxTtl = 0
            for i in range(4):
                mxTtl = max(mxTtl, dfs(r+direc[i], c+direc[i+1]))
            grid[r][c] *= -1
            return mxTtl + grid[r][c]
        
        ans = 0
        for r in range(n):
            for c in range(m):
                if grid[r][c] != 0:
                    ans = max(ans, dfs(r, c))
        return ans
            

File: 1219-path-with-maximum-gold/1219-path-with-maximum-gold.cpp
class Solution {
public:
    int getMaximumGold(vector<vector<int>>& grid) {
        int n = grid.size(), m = grid[0].size();
        int res = 0;
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++){
                if(grid[i][j] != 0){
                    res = max(res, bt(grid, i, j));
                }
            }
        } 
        return res;
    }
                              
    int bt(vector<vector<int>> &grid, int x, int y){
        int val = grid[x][y];
        grid[x][y] = 0;
        
        int ttl = 0;
        for(int d = 0; d < 4; d++){
            int i = x + dir[d];
            int j = y + dir[d+1];
            
            if(i < 0 || j < 0 || i >= grid.size() || j >= grid[0].size() || grid[i][j] == 0) {
                continue;
            }
            
            ttl = max(ttl, bt(grid, i, j));
        }
        
        grid[x][y] = val;
        return ttl + val;
    }
                              
    int dir[5] = {0, 1, 0, -1, 0};
};